homeassistant/components/osoenergy/sensor.py

```diff
@@ -0,0 +1,181 @@
+        device_class=SensorDeviceClass.ENUM,
+        value=lambda device: ENUM_VALUE_MAPPING["heater_mode"].get(
+            device.state.lower(), device.state.lower()
+        ),
+    ),
+    OSOEnergySensorEntityDescription(
+        key="optimization_mode",
+        translation_key="optimization_mode",
+        device_class=SensorDeviceClass.ENUM,
+        value=lambda device: ENUM_VALUE_MAPPING["optimization_mode"].get(
+            device.state.lower(), device.state.lower()
+        ),
+    ),
+    OSOEnergySensorEntityDescription(
+        key="power_load",
+        translation_key="power_load",
+        device_class=SensorDeviceClass.POWER,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.KILO_WATT,
+        value=lambda device: device.state,
+    ),
+    OSOEnergySensorEntityDescription(
+        key="tapping_capacity_kwh",
+        translation_key="tapping_capacity_kwh",
+        device_class=SensorDeviceClass.ENERGY,
+        native_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+        value=lambda device: device.state,
+    ),
+    OSOEnergySensorEntityDescription(
+        key="capacity_mixed_water_40",
+        translation_key="capacity_mixed_water_40",
+        device_class=SensorDeviceClass.VOLUME,
+        native_unit_of_measurement=UnitOfVolume.LITERS,
+        value=lambda device: device.state,
+    ),
+    OSOEnergySensorEntityDescription(
+        key="v40_min",
+        translation_key="v40_min",
+        device_class=SensorDeviceClass.VOLUME,
+        native_unit_of_measurement=UnitOfVolume.LITERS,
+        value=lambda device: device.state,
+    ),
+    OSOEnergySensorEntityDescription(
+        key="v40_level_min",
+        translation_key="v40_level_min",
+        device_class=SensorDeviceClass.VOLUME,
+        native_unit_of_measurement=UnitOfVolume.LITERS,
+        value=lambda device: device.state,
+    ),
+    OSOEnergySensorEntityDescription(
+        key="v40_level_max",
+        translation_key="v40_level_max",
+        device_class=SensorDeviceClass.VOLUME,
+        native_unit_of_measurement=UnitOfVolume.LITERS,
+        value=lambda device: device.state,
+    ),
+    OSOEnergySensorEntityDescription(
+        key="volume",
+        translation_key="volume",
+        device_class=SensorDeviceClass.VOLUME,
+        native_unit_of_measurement=UnitOfVolume.LITERS,
+        value=lambda device: device.state,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up OSO Energy sensor."""
+    osoenergy = hass.data[DOMAIN][entry.entry_id]
+    devices = osoenergy.session.device_list.get("sensor")
+    entities = []
+    if devices:
+        for dev in devices:
+            for description in SENSOR_TYPES:
+                if dev.osoEnergyType.lower() == description.key:
+                    entities.append(OSOEnergySensor(osoenergy, description, dev))
```

Comment:
So wait, the `dev` is now for example `v40_level_max`. And we loop over every entity description to find the right description with the key? Why don't you make the sensor_types a `dict[str, OSOEnergySensorEntityDescription]`? And have the value of the dict be the key as well and then just lookup?